core/include/gnuradio-4.0/BlockModel.hpp

```diff
@@ -546,6 +546,186 @@ class BlockWrapper : public BlockModel {
     [[nodiscard]] void*                      raw() override { return std::addressof(blockRef()); }
 };
 
+property_map serializeEdge(const auto& edge) {
+    using namespace std::string_literals;
+    property_map result;
+    auto         serializePortDefinition = [&](const std::string& key, const PortDefinition& portDefinition) {
+        std::visit(meta::overloaded( //
+                       [&](const PortDefinition::IndexBased& definition) {
+                           result[key + ".top_level"] = definition.topLevel;
+                           result[key + ".sub_index"] = definition.subIndex;
+                       }, //
+                       [&](const PortDefinition::StringBased& definition) { result[key] = definition.name; }),
+                    portDefinition.definition);
+    };
+
+    result["source_block"s] = std::string(edge.sourceBlock()->uniqueName());
+    serializePortDefinition("source_port"s, edge.sourcePortDefinition());
+    result["destination_block"s] = std::string(edge.destinationBlock()->uniqueName());
+    serializePortDefinition("destination_port"s, edge.destinationPortDefinition());
+
+    result["weight"s]          = edge.weight();
+    result["min_buffer_size"s] = edge.minBufferSize();
+    result["edge_name"s]       = std::string(edge.name());
+
+    result["buffer_size"s] = edge.bufferSize();
+    result["edge_state"s]  = std::string(magic_enum::enum_name(edge.state()));
+    result["n_readers"s]   = edge.nReaders();
+    result["n_writers"s]   = edge.nWriters();
+    result["type"s]        = std::string(magic_enum::enum_name(edge.edgeType()));
+
+    return result;
+};
+
+namespace BlockSerializationFlags {
+enum Flags : int {
+    Data       = 1,  //
+    StaticData = 2,  //
+    Ports      = 4,  //
+    Settings   = 8,  //
+    Children   = 16, //
+    All        = 0xFF
+};
+}
+
+template<typename PluginLoader>
+requires(not std::is_pointer_v<std::remove_cvref_t<PluginLoader>>)
+inline property_map serializeBlock(PluginLoader& pluginLoader, const std::shared_ptr<BlockModel>& block, int flags) {
```

Comment:
Not clear why we need this function. This is already present in the Settings implementatino.

Reply:
Does this have to be part of the BlockModel.hpp?

Reply:
> Is it urgnetly needed in the GR4 library?

It is already a part of GR4, it is moved from another place. And yes, it is needed, it is used for GRC saving and for communicating information about the flowgraph via the messaging system.

> Does it need to reside in BlockModel[.hpp]. 

It doesn't need to, can be in a separate header. But it is a function that serializes info about a BlockModel into a map. It can not go to yaml importer exporter as it is also used for messaging.